### PR TITLE
Issue 8538 19267 v2

### PR DESCRIPTION
**Pull request by erickgonzalez**
CategoryAjax: parents were not set when editing a category
CategoryAPI: when editing category check permissions over itself
PermissionAPI: new method to check permissions over a specific object type and asset.

Tests for everything.

### CODE REVIEW

**Review by dsilvam — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/business/PermissionBitAPIImpl.java [hunk 1]
```diff
@@ -1415,24 +1415,35 @@ private boolean isHost(final Permissionable permissionable) {
...
+		if(APILocator.getUserAPI().isCMSAdmin(user)) return true;
 
-    	Boolean hasPerm = false;
-    	RoleAPI roleAPI = APILocator.getRoleAPI();
-		List<com.dotmarketing.business.Role> roles = roleAPI.loadRolesForUser(user.getUserId(), false);
-		for(com.dotmarketing.business.Role r : roles) {
+		Boolean hasPerm = false;
+		final RoleAPI roleAPI = APILocator.getRoleAPI();
+		final List<com.dotmarketing.business.Role> roles = roleAPI.loadRolesForUser(user.getUserId(), false);
+		for(final com.dotmarketing.business.Role r : roles) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/PermissionBitAPIImpl.java:1430]: rename to role

File: dotCMS/src/main/java/com/dotmarketing/business/PermissionBitAPIImpl.java [hunk 2]
```diff
@@ -1415,24 +1415,35 @@ private boolean isHost(final Permissionable permissionable) {
 
 	@Override
     public boolean doesUserHavePermissions(PermissionableType permType, int permissionType, User user) throws DotDataException {
-    	if(user==null) return false;
+    	return doesUserHavePermissions(null,permType,permissionType,user);
+    }
+
+	@Override
+	public boolean doesUserHavePermissions(final String assetId, final PermissionableType permType, final int permissionType, final User user) throws DotDataException {
+		if(user==null) return false;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/PermissionBitAPIImpl.java:1423]: use curly brackets

File: dotCMS/src/main/java/com/dotmarketing/business/PermissionBitAPIImpl.java [hunk 3]
```diff
@@ -1415,24 +1415,35 @@ private boolean isHost(final Permissionable permissionable) {
...
-    	if(user==null) return false;
+    	return doesUserHavePermissions(null,permType,permissionType,user);
+    }
+
+	@Override
+	public boolean doesUserHavePermissions(final String assetId, final PermissionableType permType, final int permissionType, final User user) throws DotDataException {
+		if(user==null) return false;
 
-    	if(APILocator.getUserAPI().isCMSAdmin(user)) return true;
+		if(APILocator.getUserAPI().isCMSAdmin(user)) return true;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/PermissionBitAPIImpl.java:1425]: use curly brackets

File: dotCMS/src/main/java/com/dotmarketing/business/PermissionBitAPIImpl.java [hunk 4]
```diff
@@ -1415,24 +1415,35 @@ private boolean isHost(final Permissionable permissionable) {
...
+		for(final com.dotmarketing.business.Role r : roles) {
 			List<Permission> perms = APILocator.getPermissionAPI().getPermissionsByRole(r, false);
-			for (Permission p : perms) {
+			if(UtilMethods.isSet(assetId)) {
+				perms = perms.stream()
+						.filter(permission -> permission.getInode().equalsIgnoreCase(assetId))
+						.collect(
+								Collectors.toList());
+			}
+			for (final Permission p : perms) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/PermissionBitAPIImpl.java:1438]: rename to permission

**Review by jdotcms — APPROVED** [review]

**Review by fabrizzio-dotCMS — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/business/PermissionBitAPIImpl.java (hunk 4 above)
Comment (erickgonzalez) [reply dotCMS/src/main/java/com/dotmarketing/business/PermissionBitAPIImpl.java:1438]: done

**Pull request merged**